text/0000-reads-clauses-on-methods.md

```diff
@@ -0,0 +1,200 @@
+- Feature Name: reads_clauses_on_methods
+- Start Date: 2021-09-17
+- RFC PR: [dafny-lang/rfcs#0000](https://github.com/dafny-lang/rfcs/pull/0000)
+- Dafny Issue: [dafny-lang/dafny#0000](https://github.com/dafny-lang/dafny/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This proposal simply adds support for optional `reads` clauses on methods, just as they are already supported on functions. While the default for functions is disallowing all reads (i.e. `reads {}`) the default for methods would be allowing all reads (i.e. `reads *`), allowing full backwards-compatibility.
+
+# Motivation
+[motivation]: #motivation
+
+Dafny currently has no direct support for concurrency. The soundness of method verification depends on the assumption that no modification to mutable state occurs other than what is present in the method body. Consider this example of a Dafny class with internal state:
+
+```dafny
+class Counter {
+  var x: int
+  constructor() {
+    x := 0;
+  }
+  function method Get() 
+    reads this
+    ensures Get() == x
+  {
+    x
+  }
+  method Increment() 
+    modifies this
+    ensures x == old(x) + 1
+  {
+    x := x + 1;
+  }
+}
+```
+
+If this class is compiled to C#, for example, and its instances are shared between multiple threads, its verified properties are no longer true. If two threads both invoke `Increment` on a single counter, the effect could easily be to only increment `x` once, since reading and writing `x` does not occur atomically. Even the post-condition on `Get` is no longer valid, because the value of `x` could change in-between reading its value and returning from the compiled body.
+
+This proposal does not aim to add concurrency primitives to Dafny, or even propose any particular choice of paradigm for asynchronous or concurrent execution. Instead, it supports verifying that a particular method is safe to use in a concurrent environment, by verifying that it does not operate on any mutable state unsafely.
+
+# Guide-level explanation
```

Comment:
Other `reads` frames are necessary on internal methods (i.e. not exposed to external code) in the control flow though. For example, `FibonacciMemoizer.Get` has to declare `reads this`, because otherwise the default spec of `reads *` would violate the tighter `reads {}` frame of `SafeFirstNFibonaccis`.

I would actually argue that reusing an existing concept in a very similar additional context is a lower cognitive burden on programmers than a new concept like `threadsafe` (and again I'm trying not to introduce the concept of a "thread" unless we're 100% sure that's the right concurrency model for Dafny).

Reply:
> I would actually argue that reusing an existing concept in a very similar additional context is a lower cognitive burden on programmers than a new concept

I guess that's a difficult thing to determine, but I was thinking that if there is a specific pattern (`reads {}`) used for a specific use-case (calling externally exposed methods concurrently), it does help to name that use-case and remove the need for the pattern.

> Other reads frames are necessary on internal methods (i.e. not exposed to external code) in the control flow though. For example, FibonacciMemoizer.Get has to declare reads this, because otherwise the default spec of reads * would violate the tighter reads {} frame of SafeFirstNFibonaccis.

I see. I'm a little confused by the file `text/guide-level-explanation-example-1.dfy` which has similar but different content that's what's inline in `text/0000-reads-clauses-on-methods.md`. The former has `reads this` but the latter one doesn't. 

Given that `FibonacciMemoizer.Get` doesn't take any reference type parameters, isn't `reads this` the upper bound of what it can read so `reads this` can be implicit?

Reply:
> I guess that's a difficult thing to determine, but I was thinking that if there is a specific pattern (reads {}) used for a specific use-case (calling externally exposed methods concurrently), it does help to name that use-case and remove the need for the pattern.

That's true. I could imagine something like `{:threadsafeExtern}`, or some way of parameterizing `{:extern}`. It wouldn't eliminate the need for `reads` clauses in general in order to verify this property, though. You'd have to either have that annotation imply `reads {}` or require that such methods also declare `reads {}` explicitly, and I'd likely prefer the latter.

Given `{:extern}` is currently very weakly specified and already has several other sharp edges (https://github.com/dafny-lang/dafny/labels/foreign-function-interface), it's my preference to leave this out of scope for now, just as we also want to leave any concurrency semantics for Dafny out of scope if we can. 

> I see. I'm a little confused by the file text/guide-level-explanation-example-1.dfy which has similar but different content that's what's inline in text/0000-reads-clauses-on-methods.md. The former has reads this but the latter one doesn't.

Ah that's my fault for letting the two versions get out of sync. I was hoping to find a way to include the source files into the markdown. I've deleted them and put the missing `reads this` clause into the example. 

Some classes will only have `const` fields, so some class methods can declare `reads {}` rather than `reads this`. I'm still thinking making the default `reads *` everywhere would be better for simplicity.

Reply:
> That's true. I could imagine something like {:threadsafeExtern}, or some way of parameterizing {:extern}. It wouldn't eliminate the need for reads clauses in general in order to verify this property, though. You'd have to either have that annotation imply reads {} or require that such methods also declare reads {} explicitly, and I'd likely prefer the latter.

I think I agree that `reads` clauses on methods are the more lightweight feature to add since it doesn't introduce any new concept.

Reply:
If we're okay with a keyword like `threadsafe`, then the language could say the following:
* A method that is not declared to be `threadsafe` is not allowed to have a `reads` clause. Such a method is allowed to read anything (just like today).
* A method declared with `threadsafe method` is allowed to have a `reads` clause. We now even have the freedom to say that the default `reads` clause for a `threadsafe method` is the empty set. This would make it consistent with both `reads` and `modifies` clauses today. Furthermore, this would completely address my concern about beginning users getting the wrong idea that they must declare the `reads` set for method -- if they ever see, or try to declare, a method _with_ a `reads` clause, then that method is declared as `threadsafe method`, which the user can identify.